File: Tugas Besar 1/Facility/Truck.cpp
/**
 * @file Truck.hpp
 * @author 13515125/Muhammad Rafid Amrullah, 13517092/Michael Ray, 13517113/Juro Sutantra, 13517137/Vincent Budianto
 * @brief file berisi definisi class Truck
 * @version 0.1
 * @date 2019-03-20
 */

/* NIM/Nama  : 13515125/Muhammad Rafid Amrullah, 13517092/Michael Ray, 13517113/Juro Sutantra, 13517137/Vincent Budianto
 * Nama file : Truck.hpp
 * Topik     : Tugas Besar 1 IF2210 - Pemograman Berorientasi Objek
 * Tanggal   : 20 Maret 2019
 * Deskripsi : Header Truck.cpp */

#include <ctime>
#include <memory>

#include "Truck.hpp"

using namespace std;

/**
 * @brief Construct a new Truck object
 * 
 */
Truck::Truck(){
	setY(1);
	setX(10);
	lastTime = time(nullptr);
	firstTime = true;
}

/**
 * @brief Destroy the Truck object
 * 
 */
Truck::~Truck(){
	
}

/**
 * @brief Method to render the Facility type character to map
 * 
 * @return char 
 */
char Truck::render(){
	return 'T';
}

/**
 * @brief Get the Availability object
 * 
 * @return int 
 */
int Truck::isAvailable(){
	time_t now = time(nullptr);
	tm now_time = *localtime(addressof(now));
	tm last_time = *localtime(addressof(lastTime));

	return (now_time.tm_min - last_time.tm_min >= 1 or firstTime);
}

/**
 * @brief Set the Last Time object
 * 
 */
void Truck::setLastTime(){
	lastTime = time(nullptr);
}

/**
 * @brief Method to sell all items in inventory
 * 
 */
int Truck::transact(){
	int valid = isAvailable();
	if(valid){
		setLastTime();
		firstTime = false;
	}
	return valid;
}

File: Java/model/facility/Truck.java
/**
 * Truck.java
 * <p>
 * File berisi interface Truck
 * </p>
 * @author 13515125/Muhammad Rafid Amrullah, 13517113/Juro Sutantra, 13517137/Vincent Budianto
 * @version 2.0
 * @since 2019-04-21
 */

import java.io.*;
import java.lang.*;
import java.util.*;
import java.time.Instant;
import java.time.Duration;

/**
 * <p>
 * Class for rendering object to Truck
 * </p>
 */
public class Truck extends Facility
{
	private Instant latestTime;
	private boolean firstTime;
	/**
	 * Truck constructor
	 * @param y
	 * @param x
	 * @return Character
	 */
	public Truck(int y, int x){
		setY(y);
		setX(x);
		latestTime = Instant.now();
		firstTime = true;
	}

	/**
	 * Render Truck to map
	 * @return Character
	 */
	public Character render(){
		return 'T';
	}

	/**
	 * Check truck avaliablity
	 * @return boolean
	 */
	public boolean isAvailable(){
		Instant nowTime = Instant.now();
		Duration timeElapsed = Duration.between(latestTime, nowTime);

		return (timeElapsed.toMinutes() >= 1 || firstTime);
	}

	/**
	 * Set latestTime attribute
	 */
	public void setLastTime(){
		latestTime = Instant.now();
	}

	/**
	 * Method for trasacting
	 * @return boolean
	 */
	public boolean transact(){
		boolean valid = isAvailable();
		if(valid){
			setLastTime();
			firstTime = false;
		}
		return valid;
	}
}
